### PR TITLE
HDDS-8709. Confirm snapshot background services start up correctly after bootstrapping

### PR DESCRIPTION
**Pull request by mladjan-gadzic**
## What changes were proposed in this pull request?
Integration test that verifies background processes start correctly after leadership transfer.

## What is the link to the Apache JIRA
[https://issues.apache.org/jira/browse/HDDS-8709](https://issues.apache.org/jira/browse/HDDS-8709)

## How was this patch tested?
Integration test.

### CODE REVIEW

**Comment by prashantpogde** [comment]
These are good additional tests to have. Thank you @mladjan-gadzic  for the contribution.

**Comment by prashantpogde** [comment]
@swamirishi please take a look too.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java [hunk 1]
```diff
@@ -1036,6 +1070,304 @@ public void testInstallCorruptedCheckpointFailure() throws Exception {
...
+    // Verify RPC server is running
+    GenericTestUtils.waitFor(followerOM::isOmRpcServerRunning, 100, 5000);
+
+    // Read & Write after snapshot installed.
+    List<String> newKeys = writeKeys(1);
+    readKeys(newKeys);
+
+    checkSnapshot(leaderOM, followerOM, snapshotName, keys, snapshotInfo);
+
+    // verify that the bootstrap Follower OM can become leader again
```
Comment (GeorgeJahad) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java:1132]: It is not really "again", is it?  this is the first time it is becoming the leader, right?

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java [hunk 2]
```diff
@@ -1036,6 +1070,304 @@ public void testInstallCorruptedCheckpointFailure() throws Exception {
...
+    Assertions.assertNotNull(sstBackupDir);
+    Path sstBackupDirPath = Paths.get(sstBackupDir);
+    int numberOfSstFiles = 0;
+    try (DirectoryStream<Path> files =
+             Files.newDirectoryStream(sstBackupDirPath)) {
+      for (Path ignored : files) {
+        numberOfSstFiles++;
+      }
+    }
+
```
Comment (GeorgeJahad) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java:1167]: I'd like an assert that numberOfSstFiles is > 0, as a safety net.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java (hunk 1 above)
Comment (mladjan-gadzic) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java]: You are right. I fixed a comment so it does not introduce confusion.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java [hunk 3]
```diff
@@ -1036,6 +1070,304 @@ public void testInstallCorruptedCheckpointFailure() throws Exception {
...
+      GenericTestUtils.waitFor(() -> {
+        long newUsedBytes;
+        try {
+          newUsedBytes = getUsedBytes(newLeaderOM);
+        } catch (IOException e) {
+          throw new RuntimeException(e);
+        }
+        return newUsedBytes == usedBytes - newKeyInfo.getDataSize();
+      }, 1000, 10000);
+    } catch (TimeoutException e) {
```
Comment (GeorgeJahad) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java:1243]: I don't understand this catch clause; it seems to be forcing the very thing we are trying to test.  Am I missing something?

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java [hunk 4]
```diff
@@ -1036,6 +1069,292 @@ public void testInstallCorruptedCheckpointFailure() throws Exception {
...
+    Assertions.assertNotNull(newKeyInfo);
+
+    long usedBytes = getUsedBytes(newLeaderOM);
+
+    ozoneBucket.deleteKeys(newKeys);
+
+    GenericTestUtils.waitFor(() -> {
+      try {
+        return getUsedBytes(newLeaderOM) ==
+            usedBytes - newKeyInfo.getDataSize();
```
Comment (GeorgeJahad) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java:1232]: newKey is in the newSnapshot right?  So then this test actually should fail, because the snapshot still exists. Something isn't right.  Please take a look.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java [hunk 5]
```diff
@@ -1036,6 +1069,292 @@ public void testInstallCorruptedCheckpointFailure() throws Exception {
...
+      } catch (IOException e) {
+        throw new RuntimeException(e);
```
Comment (GeorgeJahad) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java:1234]: throwing to show a test failure is not really the convention.  better to use Assert.fail()

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java [hunk 6]
```diff
@@ -1036,6 +1069,292 @@ public void testInstallCorruptedCheckpointFailure() throws Exception {
...
+      }
+    }, 1000, 10000);
+
+    GenericTestUtils.waitFor(() -> {
+      Table<String, OmKeyInfo> omKeyInfoTableAfterDeletion = newLeaderOM
```
Comment (GeorgeJahad) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java:1239]: What is the reason for creating a new variable here: omKeyInfoTableAfterDeletion  Isn't it going to be identical to omKeyInfoTableBeforeDeletion?

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java (hunk 4 above)
Comment (mladjan-gadzic) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java:1232]: > newKey is in the newSnapshot right?

That is correct.

My understanding is that only key gets deleted here. Did I misunderstand something?

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java [hunk 7]
```diff
@@ -1036,6 +1069,292 @@ public void testInstallCorruptedCheckpointFailure() throws Exception {
...
+    OzoneManager newLeaderOM = cluster.getOMLeader();
+    Assertions.assertEquals(followerOM, newLeaderOM);
+    OzoneManager newFollowerOM =
+        cluster.getOzoneManager(leaderOM.getOMNodeId());
+    Assertions.assertEquals(leaderOM, newFollowerOM);
+
+    checkSnapshot(newLeaderOM, newFollowerOM, snapshotName, keys, snapshotInfo);
+    readKeys(newKeys);
+
+    // Prepare baseline data for SST pruning
```
Comment (GeorgeJahad) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java:1151]: For these "pruning tests" lets change the comments to call it "compaction backup pruning" instead "sst pruning" to make it very clear you are not referring to snapshot filtering service.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java [hunk 8]
```diff
@@ -1036,6 +1069,292 @@ public void testInstallCorruptedCheckpointFailure() throws Exception {
...
+          .mapToLong(String::length)
+          .reduce(0L, Long::sum);
+      for (Path ignored : files) {
+        newNumberOfLogFiles++;
+      }
+    }
+    Assertions.assertTrue(numberOfLogFiles != newNumberOfLogFiles
+        || contentLength != newContentLength);
+
+    // Check whether sst files were pruned
```
Comment (GeorgeJahad) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java:1283]: "compaction backup files were pruned"

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java [hunk 9]
```diff
@@ -1036,6 +1069,292 @@ public void testInstallCorruptedCheckpointFailure() throws Exception {
...
+                 Files.newDirectoryStream(sstBackupDirPath)) {
+          for (Path ignored : files) {
+            newNumberOfSstFiles++;
+          }
+        } catch (IOException e) {
+          throw new RuntimeException(e);
+        }
+        return finalNumberOfSstFiles > newNumberOfSstFiles;
+      }, 1000, 10000);
+    } catch (TimeoutException e) {
```
Comment (GeorgeJahad) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java:1298]: like the one above, this catch clause seems to defeat the purpose of the test.  Is the test flaky without the catch clause?

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java [hunk 10]
```diff
@@ -1036,6 +1069,292 @@ public void testInstallCorruptedCheckpointFailure() throws Exception {
...
+          newLeaderOM.getMetadataManager().getSnapshotInfoTable();
+      try {
+        return null == snapshotInfoTable.get(newSnapshot.getTableKey());
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+    }, 1000, 10000);
+
+    // Check whether compaction logs get appeneded to
+    // Force compaction
```
Comment (GeorgeJahad) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java:1263]: "by forcing compaction"

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java (hunk 9 above)
Comment (GeorgeJahad) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java:1298]: (Please run the 100 iterations in the ci, like we did earlier.)

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java (hunk 3 above)
Comment (mladjan-gadzic) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java]: It was a leftover during testing. Removed.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java (hunk 9 above)
Comment (mladjan-gadzic) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java:1298]: This one is flaky without it. The reason is that it is hard to set interval of pruning correctly because tests before does not execute in constant time.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java (hunk 10 above)
Comment (mladjan-gadzic) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java:1263]: Done.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java (hunk 6 above)
Comment (mladjan-gadzic) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java:1239]: TBH not sure why, removed.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java (hunk 5 above)
Comment (mladjan-gadzic) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java:1234]: Fixed.

File: .github/workflows/ci.yml
Comment (smengcl) [inline .github/workflows/ci.yml:1]: Changes in this file and `junit.sh` should be in a separate dev branch and does not belong to this PR?

Comment (mladjan-gadzic) [reply .github/workflows/ci.yml:1]: You are right, I will revert those and move to separate dev branch.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java (hunk 9 above)
Comment (mladjan-gadzic) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java:1295]: As per [smenglc comment](https://github.com/apache/ozone/pull/5083#discussion_r1270234790) I moved this change to separate dev branch [https://github.com/mladjan-gadzic/ozone/tree/HDDS-8709-4](https://github.com/mladjan-gadzic/ozone/tree/HDDS-8709-4).

**Review by hemantk-12 — COMMENTED** [review]
Thanks for adding new tests @mladjan-gadzic . This is good addition to the CD/CI.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java [hunk 11]
```diff
@@ -143,6 +162,20 @@ public void init(TestInfo testInfo) throws Exception {
         StorageUnit.KB);
     conf.setStorageSize(OMConfigKeys.
         OZONE_OM_RATIS_SEGMENT_PREALLOCATED_SIZE_KEY, 16, StorageUnit.KB);
+    if (testInfo.getDisplayName().equals("testSnapshotBackgroundServices")) {
+      conf.setTimeDuration(OZONE_SNAPSHOT_SST_FILTERING_SERVICE_INTERVAL,
+          5, TimeUnit.SECONDS);
+      conf.setTimeDuration(OZONE_SNAPSHOT_DELETING_SERVICE_INTERVAL, 5,
+          TimeUnit.SECONDS);
+      conf.setTimeDuration(OZONE_OM_SNAPSHOT_COMPACTION_DAG_MAX_TIME_ALLOWED,
```
[the remark this answers is not in the document]
Comment (mladjan-gadzic) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java:170]: Compaction log pruning is being tested as well, that is why this is necessary.

**Comment by mladjan-gadzic** [comment]
> Thanks for adding new tests @mladjan-gadzic . This is good addition to the CD/CI.

Thank you @hemantk-12 for having a look at it!

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java [hunk 12]
```diff
@@ -1036,6 +1069,272 @@ public void testInstallCorruptedCheckpointFailure() throws Exception {
...
+      Table<String, SnapshotInfo> snapshotInfoTable =
+          newLeaderOM.getMetadataManager().getSnapshotInfoTable();
+      try {
+        return null == snapshotInfoTable.get(newSnapshot.getTableKey());
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+    }, 1000, 10000);
+
+    // Check whether compaction logs get appeneded to by forcning compaction
```
Comment (hemantk-12) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java:1260]:
```suggestion
    // Check whether compaction logs get appended to by forcing compaction
```

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java [hunk 13]
```diff
@@ -1036,6 +1069,272 @@ public void testInstallCorruptedCheckpointFailure() throws Exception {
...
+    Assertions.assertEquals(followerOM, newLeaderOM);
+    OzoneManager newFollowerOM =
+        cluster.getOzoneManager(leaderOM.getOMNodeId());
+    Assertions.assertEquals(leaderOM, newFollowerOM);
+
+    checkSnapshot(newLeaderOM, newFollowerOM, snapshotName, keys, snapshotInfo);
+    readKeys(newKeys);
+
+    // Prepare baseline data for compaction logs
+    String currentCompactionLogPath = newLeaderOM
```
Comment (hemantk-12) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java:1168]: Shouldn't this be after line 1117 just after you are doing baseline of SST files in backDir?

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java [hunk 14]
```diff
@@ -1036,6 +1069,272 @@ public void testInstallCorruptedCheckpointFailure() throws Exception {
...
+        snapshotNamePrefix + RandomStringUtils.randomNumeric(10)).getName();
+    SnapshotDiffReportOzone diff = getSnapDiffReport(volumeName, bucketName,
+        firstSnapshot, secondSnapshot);
+    Assertions.assertEquals(Collections.singletonList(
+            SnapshotDiffReportOzone.getDiffReportEntry(
+                SnapshotDiffReport.DiffType.CREATE, diffKey, null)),
+        diff.getDiffList());
+  }
+
+  private long getUsedBytes(OzoneManager newLeaderOM) throws IOException {
```
Comment (hemantk-12) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java:1309]: nit: just call it `leaderOm` or `ozoneManager`.
```suggestion
  private long getUsedBytes(OzoneManager leaderOm) throws IOException {
```

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java [hunk 15]
```diff
@@ -1036,6 +1069,272 @@ public void testInstallCorruptedCheckpointFailure() throws Exception {
...
+    return newLeaderOM
+        .getBucketManager()
+        .getBucketInfo(ozoneBucket.getVolumeName(), ozoneBucket.getName())
+        .getUsedBytes();
+  }
+
+  private SnapshotDiffReportOzone getSnapDiffReport(String volume,
```
Comment (hemantk-12) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java:1316]: nit: alignment is bit off.

```suggestion
  private SnapshotDiffReportOzone getSnapDiffReport(String volume,
                                                    String bucket,
                                                    String fromSnapshot,
                                                    String toSnapshot)
      throws InterruptedException, TimeoutException {
```

Or 

```
  private SnapshotDiffReportOzone getSnapDiffReport(
      String volume,
      String bucket,
      String fromSnapshot,
      String toSnapshot
  ) throws InterruptedException, TimeoutException {
```

FYI, you can setup [Ozone style](https://github.com/apache/ozone/blob/master/hadoop-ozone/dev-support/intellij/ozone-style.xml) in IDE.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java [hunk 16]
```diff
@@ -1036,6 +1069,272 @@ public void testInstallCorruptedCheckpointFailure() throws Exception {
...
+      try {
+        response.set(client.getObjectStore()
+            .snapshotDiff(
+                volume, bucket, fromSnapshot, toSnapshot, null, 0, false,
+                false));
+        return response.get().getJobStatus() == DONE;
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+    }, 1000, 10000);
```
Comment (hemantk-12) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java:1333]: nit: You can also use `SnapshotDiffResponse`'s waitTime for polling.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java [hunk 17]
```diff
@@ -1036,6 +1069,272 @@ public void testInstallCorruptedCheckpointFailure() throws Exception {
...
+      return finalNumberOfSstFiles > newNumberOfSstFiles;
+    }, 1000, 10000);
+
+    // Snap diff
+    String firstSnapshot = createOzoneSnapshot(newLeaderOM,
+        snapshotNamePrefix + RandomStringUtils.randomNumeric(10)).getName();
+    String diffKey = writeKeys(1).get(0);
+    String secondSnapshot = createOzoneSnapshot(newLeaderOM,
+        snapshotNamePrefix + RandomStringUtils.randomNumeric(10)).getName();
+    SnapshotDiffReportOzone diff = getSnapDiffReport(volumeName, bucketName,
```
Comment (hemantk-12) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java:1301]: Do we also need to check the correctness of snapshots created before the bootstrapping and transfer of leadership? I don't see we are doing any validation of the SST files as such. Please correct me if I'm wrong.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java [hunk 18]
```diff
@@ -1036,6 +1074,349 @@ public void testInstallCorruptedCheckpointFailure() throws Exception {
     assertLogCapture(logCapture, msg);
   }
 
+  @Test
+  @DisplayName("testSnapshotBackgroundServices")
+  @SuppressWarnings("methodlength")
+  public void testSnapshotBackgroundServices()
+      throws Exception {
```
Comment (smengcl) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java:1081]: It's a 300+-line test. It might be worth to write a block of comment at the beginning to state what is being tested.

**Review by smengcl — COMMENTED** [review]
Thanks @mladjan-gadzic a lot for the test addition. I don't have much to add. Pls address the remaining comments and it can be merged.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java [hunk 19]
```diff
@@ -1036,6 +1069,272 @@ public void testInstallCorruptedCheckpointFailure() throws Exception {
...
+    cluster.startInactiveOM(followerNodeId);
+
+    // The recently started OM should be lagging behind the leader OM.
+    // Wait & for follower to update transactions to leader snapshot index.
+    // Timeout error if follower does not load update within 10s
+    GenericTestUtils.waitFor(() ->
+        followerOM.getOmRatisServer().getLastAppliedTermIndex().getIndex()
+            >= leaderOMSnapshotIndex - 1, 100, 10000);
+
+
```
[the remark this answers is not in the document]
Comment (mladjan-gadzic) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java]: Done.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java (hunk 12 above)
Comment (mladjan-gadzic) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java]: Fixed.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java (hunk 13 above)
Comment (mladjan-gadzic) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java]: Fixed.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java (hunk 15 above)
Comment (mladjan-gadzic) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java:1408]: Thanks! Somehow I missed this when I was onboarded on project.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java (hunk 16 above)
Comment (mladjan-gadzic) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java]: Done.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java (hunk 17 above)
Comment (mladjan-gadzic) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java:1400]: Goal of this test is only to check whether background services (SST filtering, key/snapshot deletion and compaction backup pruning) work after leadership transfer. My understanding is that what we did is sufficient for described case.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java (hunk 18 above)
Comment (mladjan-gadzic) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java:1092]: Done.

**Comment by mladjan-gadzic** [comment]
> Thanks @mladjan-gadzic a lot for the test addition. I don't have much to add. Pls address the remaining comments and it can be merged.

Thanks @smengcl for a review!

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java (hunk 17 above)
Comment (hemantk-12) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java:1400]: It is OK to have this test validate the correctness of system's state. But I think we should also have a test to verify the correctness if we don't have that and it doesn't need to be part of this change.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java [hunk 20]
```diff
@@ -1036,6 +1075,359 @@ public void testInstallCorruptedCheckpointFailure() throws Exception {
...
+        }
+
+        return false;
+      } catch (IOException e) {
+        Assertions.fail();
+        return false;
+      }
+    }, 1000, 120000);
+
+    // Check whether newly created snapshot data is reclaimed
```
Comment (GeorgeJahad) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java:1346]: "reclaimed" generally means the data has been released from the data node.  That isn't really what is being tested here, is it?  How about this instead?

```suggestion
    // Confirm entry for deleted snapshot removed from info table
```

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java [hunk 21]
```diff
@@ -1036,6 +1075,359 @@ public void testInstallCorruptedCheckpointFailure() throws Exception {
...
+             Files.newDirectoryStream(compactionLogPath)) {
+      newContentLength = bufferedReader.lines()
+          .mapToLong(String::length)
+          .reduce(0L, Long::sum);
+      for (Path ignored : files) {
+        newNumberOfLogFiles++;
+      }
+    }
+    Assertions.assertTrue(numberOfLogFiles != newNumberOfLogFiles
+        || contentLength != newContentLength);
```
Comment (GeorgeJahad) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java:1377]: Can these be ">" instead of "!="?

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java [hunk 22]
```diff
@@ -1036,6 +1075,359 @@ public void testInstallCorruptedCheckpointFailure() throws Exception {
...
+      int newNumberOfSstFiles = 0;
+      try (DirectoryStream<Path> files =
+               Files.newDirectoryStream(sstBackupDirPath)) {
+        for (Path ignored : files) {
+          newNumberOfSstFiles++;
+        }
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+      return finalNumberOfSstFiles > newNumberOfSstFiles;
```
Comment (GeorgeJahad) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java:1391]: I find "finalNumberOfSstFiles" to be a confusing name.  "initialNumberOfSstFiles" seems clearer to me.

Comment (GeorgeJahad) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java:1391]: ah, I see the variable is "final" so it's got "final" in the name.  I understand your reasoning but I would still change the name.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java [hunk 23]
```diff
@@ -1036,6 +1075,359 @@ public void testInstallCorruptedCheckpointFailure() throws Exception {
...
+    }, 1000, 10000);
+
+    // Snap diff
```
Comment (GeorgeJahad) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java:1394]:
```suggestion
    // Confirm snap diff by creating 2 snapshots differing by a single key
```

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java (hunk 17 above)
Comment (mladjan-gadzic) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java:1400]: I agree. Maybe we could open Jira for this to track work and effort for that?

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java (hunk 20 above)
Comment (mladjan-gadzic) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java:1346]: Fixed.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java (hunk 22 above)
Comment (mladjan-gadzic) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java:1391]: Done.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java (hunk 17 above)
Comment (hemantk-12) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java:1383]: Created jira: https://issues.apache.org/jira/browse/HDDS-9121

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java [hunk 24]
```diff
@@ -1036,6 +1074,343 @@ public void testInstallCorruptedCheckpointFailure() throws Exception {
...
+    OzoneManager newLeaderOM = cluster.getOMLeader();
+    Assertions.assertEquals(followerOM, newLeaderOM);
+    OzoneManager newFollowerOM =
+        cluster.getOzoneManager(leaderOM.getOMNodeId());
+    Assertions.assertEquals(leaderOM, newFollowerOM);
+
+    readKeys(newKeys);
+
+    // Prepare baseline data for compaction backup pruning
+    // TODO: uncomment this when HDDS-9118 is resolved
```
Comment (hemantk-12) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java:1165]: HDDS-9118 has been merged. Please un-comment this code.

**Review by hemantk-12 — APPROVED** [review]
Thanks for the patch @mladjan-gadzic. It is great addition to the CD/CI.

LGTM.

One thing as @smengcl mentioned [here](https://github.com/apache/ozone/pull/5083#discussion_r1280535927), we should try to make components/blocks for lengthy code like this. Using blocks improves the readability as well as you can get rid of comments stating the obvious IMO.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java (hunk 24 above)
Comment (mladjan-gadzic) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMRatisSnapshots.java]: Done.

**Comment by mladjan-gadzic** [comment]
> Thanks for the patch @mladjan-gadzic. It is great addition to the CD/CI.
> 
> LGTM.
> 
> One suggestion as @smengcl mentioned [here](https://github.com/apache/ozone/pull/5083#discussion_r1280535927), we should try to make components/blocks for lengthy code like this. Using blocks improves the readability as well as you can get rid of comments stating the obvious IMO.

Thanks @hemantk-12 for taking the time to review it! I just pushed changes which should improve readability.

**Review by hemantk-12 — COMMENTED** [review]
LGTM.

**Review by smengcl — APPROVED** [review]
+1 penging CI

**Comment by smengcl** [comment]
Hi @mladjan-gadzic , it looks like the latest master branch no longer has:

```
OzoneConsts.FILTERED_SNAPSHOTS
```

which is causing the build failure. Would you take a look at this?

**Comment by mladjan-gadzic** [comment]
> Hi @mladjan-gadzic , it looks like the latest master branch no longer has:
> 
> ```
> OzoneConsts.FILTERED_SNAPSHOTS
> ```
> 
> which is causing the build failure. Would you take a look at this?

It should be fixed now. [CI](https://github.com/mladjan-gadzic/ozone/actions/runs/5881327329) passed.

**Comment by smengcl** [comment]
Thanks @mladjan-gadzic for the test addition. Thanks @prashantpogde @GeorgeJahad @hemantk-12 for reviewing this.
